app/main/views/projects.py

```diff
@@ -0,0 +1,128 @@
+from flask import jsonify, abort, request, current_app
+from sqlalchemy.exc import IntegrityError
+
+from app.main import main
+from app.models import db, Project, AuditEvent
+from app.utils import (
+    get_json_from_request, json_has_required_keys, get_int_or_400,
+    pagination_links, get_valid_page_or_1, url_for,
+    get_positive_int_or_400, validate_and_return_updater_request
+)
+
+from app.service_utils import validate_and_return_supplier
+
+from dmapiclient.audit import AuditTypes
+
+
+def get_project_json():
+    json_payload = get_json_from_request()
+    json_has_required_keys(json_payload, ['project'])
+    return json_payload['project']
+
+
+def save_project(project):
+    db.session.add(project)
+
+    try:
+        db.session.flush()
+    except IntegrityError as e:
```

Comment:
I keep cutting pasting the same code myself and wonder if it's needed. From: http://docs.sqlalchemy.org/en/latest/orm/session_basics.html
> When a flush() fails, typically for reasons like primary key, foreign key, or “not nullable” constraint violations, a rollback() is issued automatically (it’s currently not possible for a flush to continue after a partial failure). 

And
> commit...always issues flush() beforehand to flush any remaining state to the database

So I wonder if these could all become
`db.session.add()`
`db.session.commit()`